File: packages/java/src/main/java/dev/nebulatoken/MemoryRefreshTokenStore.java
package dev.nebulatoken;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Reference in-memory store ([N-21]).
 *
 * <p><b>NOT FOR PRODUCTION.</b> State is per-process and lost on restart, so
 * reuse detection does not survive a deploy and does not work behind more than
 * one instance -- the two situations in which you most want it. Implement
 * {@link RefreshTokenStore} over your database instead; see
 * {@code examples/JdbcRefreshTokenStore.java} and {@code docs/STORE.md}.
 *
 * <p>Safe for concurrent use. {@link TokenRecord} is immutable and every
 * mutation goes through {@link ConcurrentHashMap#compute} on a single key, so
 * the compare-and-set methods really are atomic rather than
 * read-then-write-and-hope -- which is what makes this store usable as the
 * oracle for the concurrency tests.
 */
public final class MemoryRefreshTokenStore implements RefreshTokenStore {

    private final Map<String, TokenRecord> rows = new ConcurrentHashMap<>();

    @Override
    public TokenRecord findBySelector(String selector) {
        return selector == null ? null : rows.get(selector);
    }

    @Override
    public void insert(TokenRecord row) {
        if (rows.putIfAbsent(row.selector(), row) != null) {
            // A selector collision means the CSPRNG repeated 16 bytes, or a
            // caller re-inserted a record. Both are infrastructure-class faults:
            // throw rather than silently overwriting a live session ([N-20]).
            throw new IllegalStateException("[NEBULA] duplicate selector " + row.selector());
        }
    }

    @Override
    public boolean markRotated(String selector, TokenStatus fromStatus, long rotatedAt,
                               String replacedBySelector) {
        AtomicBoolean applied = new AtomicBoolean(false);
        rows.computeIfPresent(selector, (key, current) -> {
            if (current.status() != fromStatus) return current; // CAS lost ([N-17])
            applied.set(true);
            return current.rotated(rotatedAt, replacedBySelector);
        });
        return applied.get();
    }

    @Override
    public boolean revokeIfActive(String selector) {
        AtomicBoolean applied = new AtomicBoolean(false);
        rows.computeIfPresent(selector, (key, current) -> {
            if (current.status() != TokenStatus.ACTIVE) return current; // CAS lost ([N-18])
            applied.set(true);
            return current.revoked();
        });
        return applied.get();
    }

    @Override
    public int revokeFamily(String familyId) {
        return revokeWhere(row -> row.familyId().equals(familyId));
    }

    @Override
    public int revokeUser(String userId) {
        return revokeWhere(row -> row.userId().equals(userId));
    }

    /** Test helper: every record currently stored. Not part of the store contract. */
    public List<TokenRecord> all() {
        return new ArrayList<>(rows.values());
    }

    /**
     * Test helper: drop records whose family deadline has passed. Records MUST
     * NOT be dropped before {@code familyExpiresAt} ([N-15]).
     */
    public int deleteExpired(long now) {
        int n = 0;
        for (Map.Entry<String, TokenRecord> entry : rows.entrySet()) {
            if (now >= entry.getValue().familyExpiresAt() && rows.remove(entry.getKey(), entry.getValue())) {
                n++;
            }
        }
        return n;
    }

    /**
     * Already-revoked rows are not counted, which is what makes revocation
     * idempotent in the sense [N-19] requires: the second call changes nothing
     * and reports 0.
     */
    private int revokeWhere(java.util.function.Predicate<TokenRecord> match) {
        int n = 0;
        for (String selector : rows.keySet()) {
            AtomicBoolean changed = new AtomicBoolean(false);
            rows.computeIfPresent(selector, (key, current) -> {
                if (current.status() == TokenStatus.REVOKED || !match.test(current)) return current;
                changed.set(true);
                return current.revoked();
            });
            if (changed.get()) n++;
        }
        return n;
    }
}

File: packages/php/src/MemoryRefreshTokenStore.php
<?php

declare(strict_types=1);

namespace NebulaToken;

/**
 * Reference store ([N-21]).
 *
 * A PHP request has a single thread of execution and none of the methods below
 * yields, so each compare-and-set is atomic with respect to the others within
 * one process.
 *
 * NOT FOR PRODUCTION: state lives in one PHP process and dies with the request,
 * so reuse detection does not survive the response, let alone a second FPM
 * worker. Implement {@see RefreshTokenStore} over your database instead — see
 * examples/PdoRefreshTokenStore.php and docs/STORE.md.
 */
final class MemoryRefreshTokenStore implements RefreshTokenStore
{
    /** @var array<string, TokenRecord> */
    private array $rows = [];

    public function findBySelector(string $selector): ?TokenRecord
    {
        return $this->rows[$selector] ?? null;
    }

    public function insert(TokenRecord $record): void
    {
        if (isset($this->rows[$record->selector])) {
            // A duplicate selector means the CSPRNG repeated itself or a caller
            // re-inserted a row: refuse rather than overwrite live state.
            // A dedicated type, not a bare RuntimeException: this is the [N-20]
            // native error channel, and a caller that wants to distinguish a
            // store refusal from any other runtime fault must be able to.
            throw new NebulaStoreException('duplicate selector');
        }
        $this->rows[$record->selector] = $record;
    }

    public function markRotated(
        string $selector,
        TokenStatus $fromStatus,
        int $rotatedAt,
        string $replacedBySelector,
    ): bool {
        $row = $this->rows[$selector] ?? null;
        if ($row === null || $row->status !== $fromStatus) {
            return false;
        }
        $this->rows[$selector] = $row->asRotated($rotatedAt, $replacedBySelector);

        return true;
    }

    public function revokeIfActive(string $selector): bool
    {
        $row = $this->rows[$selector] ?? null;
        if ($row === null || $row->status !== TokenStatus::Active) {
            return false;
        }
        $this->rows[$selector] = $row->asRevoked();

        return true;
    }

    public function revokeFamily(string $familyId): int
    {
        return $this->revokeWhere(static fn (TokenRecord $r): bool => $r->familyId === $familyId);
    }

    public function revokeUser(string $userId): int
    {
        return $this->revokeWhere(static fn (TokenRecord $r): bool => $r->userId === $userId);
    }

    /**
     * Test helper: every record currently stored. Not part of the store contract.
     *
     * @return list<TokenRecord>
     */
    public function all(): array
    {
        return array_values($this->rows);
    }

    /**
     * Test helper: drop records whose family deadline has passed ([N-15]).
     * Rotated and revoked rows before that point are what reuse detection reads.
     */
    public function deleteExpired(int $now): int
    {
        $n = 0;
        foreach ($this->rows as $selector => $row) {
            if ($now >= $row->familyExpiresAt) {
                unset($this->rows[$selector]);
                $n++;
            }
        }

        return $n;
    }

    /** @param callable(TokenRecord): bool $matches */
    private function revokeWhere(callable $matches): int
    {
        $n = 0;
        foreach ($this->rows as $selector => $row) {
            // Already-revoked rows are not counted: [N-19] asks for the number
            // of records changed, which is what makes the operation idempotent.
            if ($matches($row) && $row->status !== TokenStatus::Revoked) {
                $this->rows[$selector] = $row->asRevoked();
                $n++;
            }
        }

        return $n;
    }
}
